### PR TITLE
Fixed naming of acronyms for fields

### LINKED ISSUE #1863: Automatic naming of uppercase acronyms
The automatic naming for acronyms seems to be in an unexpected format, in my opinion.

Consider the following POCO:

```cs
public class Car
{
    public string EAN { get; set; }
}
```

The schema generated for this POCO is as below.

```graphql
type Car {
    eAN: String!
}
```

I don't know how the automatic naming works for fields but just `eAN` looks weird :) The best naming would probably be `ean`.

I will go ahead and use the `GraphQLName` attribute to fix the issue on my behalf but thought opening an issue for further discussion might be a good idea.

### PR DESCRIPTION
**Pull request by Nivisan-Jeyananthan**
Fixes acronym fieldnaming format

Previously fieldnames beginning with acronyms were formatted like : 
- GetIOFile --> iOFile
- EAN --> eAN

After this fix they would be formatted like :
- GetIOFile --> ioFile
- EAN --> ean

Closes #1863

### CODE REVIEW

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs [hunk 1]
```diff
@@ -341,8 +341,17 @@ public NameString FormatFieldName(string fieldName)
                 nameof(fieldName));
         }
 
-        return fieldName.Length > 1
-            ? fieldName.Substring(0, 1).ToLowerInvariant() + fieldName.Substring(1)
-            : fieldName.ToLowerInvariant();
+        if(fieldName.Length < 2)
+            return fieldName.ToLowerInvariant();
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:345]: can you add braces to the statements?

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs [hunk 2]
```diff
@@ -341,8 +341,17 @@ public NameString FormatFieldName(string fieldName)
...
+
+        var lastUpperCaseIndex = 0;
+
+        for (var i = 0; i < fieldName.Length; i++)
+        {
+            if (char.IsUpper(fieldName[i]))
+                lastUpperCaseIndex = i;
+        }
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:353]: I dont think this works.

Woudnt this also convert FooBarBaz to foobarBaz?

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs [hunk 3]
```diff
@@ -341,8 +341,17 @@ public NameString FormatFieldName(string fieldName)
...
+
+        return fieldName[..lastUpperCaseIndex].ToLowerInvariant() + fieldName[lastUpperCaseIndex..];
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:355]: @michaelstaib  did we already drop support for older versions? so can we use ranges now?

File: src/HotChocolate/Core/test/Types.Tests/Types/Descriptors/Conventions/DefaultNamingConventionsTypes.cs [hunk 4]
```diff
@@ -5,6 +5,24 @@ namespace HotChocolate.Types.Descriptors
 {
     public class DefaultNamingConventionsTests
     {
+
+        [InlineData("IOFile", "ioFile")]
+        [InlineData("FooBar", "fooBar")]
+        [Theory]
+        public void GetFormattedFieldName_ReturnsFormattedFieldName(string fieldName,string excpectedFieldName) {
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Tests/Types/Descriptors/Conventions/DefaultNamingConventionsTypes.cs:12]: We do have a 100 character max line length
```suggestion
        public void GetFormattedFieldName_ReturnsFormattedFieldName(
              string fieldName,
              string excpectedFieldName) {
```

File: src/HotChocolate/Core/test/Types.Tests/Types/Descriptors/Conventions/DefaultNamingConventionsTypes.cs [hunk 5]
```diff
@@ -5,6 +5,24 @@ namespace HotChocolate.Types.Descriptors
 {
     public class DefaultNamingConventionsTests
     {
+
+        [InlineData("IOFile", "ioFile")]
+        [InlineData("FooBar", "fooBar")]
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Tests/Types/Descriptors/Conventions/DefaultNamingConventionsTypes.cs:10]:
```suggestion
        [InlineData("FooBar", "fooBar")]
        [InlineData("FOOBarBaz", "fooBarBar")]
```

**Review by PascalSenn — CHANGES_REQUESTED** [review]
Couple of comments :) 

Also can you send us the errors you got when running the tests?

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 2 above)
Comment (Nivisan-Jeyananthan) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:353]: tested it in a playground and it worked

Comment (Nivisan-Jeyananthan) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:353]: but yes you are right

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 3 above)
Comment (Nivisan-Jeyananthan) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:355]: Sorry my bad, didn't consider that ppl were using still Standard 2.0 and earlier versions of c#

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 2 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:353]: @Nivisan-Jeyananthan you can try something like this i guess (might be off by one ;D )
```csharp 
var lastUpperCaseIndex = -1; 

while(lastUperCaseIndex + 1 < fieldName.Lengh && 
     char.IsUpper(fieldName[lastUpserCaseIndex+1]))  
{
    lastUpperCaseIndex++;
}
if(lastUpperCaseIndex == -1) 
{
   return fieldName;
}
return fieldName.Substring(0, lastUpperCaseIndex).ToLowerInvariant() + 
  fieldName.Substring(lastUpperCaseIndex)

```

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 3 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:355]: @Nivisan-Jeyananthan  i know, thats the sad part about developing a framework :D 

So basically there are always to different kind of new language features, Compile TIme and Runtime Features.
We can use Compile Time features immediately as these are compiled to compatible code. We can only use runtime features after we drop support for the old frameworks, what basically means, after 2 years </3

Comment (Nivisan-Jeyananthan) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:355]: @PascalSenn wow, makes sense. Thanks for telling me! :)
I was wondering why so many new C# features weren't used in your codebase.

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 2 above)
Comment (Nivisan-Jeyananthan) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:353]: @PascalSenn you mean you're a few characters short ;)

**Comment by Nivisan-Jeyananthan** [comment]
All tests will pass but the coding style may not be like you wish.
Would therefor be glad if you could suggest changes or change it to your liking.

**Comment by Nivisan-Jeyananthan** [comment]
> Couple of comments :)
> 
> Also can you send us the errors you got when running the tests?

The error I got were due to an older dotnet sdk version. 
Was trying to run stuff on dotnet 6.0.101? and the globaljson required 6.0.102?

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs [hunk 6]
```diff
@@ -332,17 +334,67 @@ public virtual bool IsDeprecated(object value, out string? reason)
     }
 
     /// <inheritdoc />
-    public NameString FormatFieldName(string fieldName)
+    public unsafe NameString FormatFieldName(string name)
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:337]: @michaelstaib  does this work for you? 
This is still a single iteration over the name but split into two loops. I do have an implementation with just one `for` but it actually is a lot more comprehesible this way

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**